File: apps/candice-companion/src/runtime/speech-timing.ts
/**
 * TTS speech-timing channel (FIX-016).
 *
 * The native shell emits three timing facts over the existing bridge
 * event surface — `candice:speech-start` (phoneme timings),
 * `candice:speech-boundary` (utterance replaced/interrupted), and
 * `candice:speech-drain` (output provably silent) — and this module
 * validates them and feeds the WS-12 `VisemeScheduler`. The scheduler
 * stays a pure, asset-free state machine; this channel is the only
 * production consumer of its timing input.
 *
 * No new auth surface: the events ride the same `app.emit` surface as
 * the authenticated bridge events, and the invoke commands live in the
 * same handler list. Validation here is format sanitation, not
 * authorization — a webview caller is already inside the shell.
 *
 * Determinism rules (inherited from WS-08/WS-12): no timers inside.
 * The channel never reads a clock of its own; the scheduler owns the
 * timing clock and this channel only feeds it validated events.
 */

import type { VisemeScheduler } from '../animation/viseme/scheduler.ts';

export const SPEECH_TIMING_SCHEMA_VERSION = '1.0';
export const SPEECH_START_EVENT = 'candice:speech-start';
export const SPEECH_BOUNDARY_EVENT = 'candice:speech-boundary';
export const SPEECH_DRAIN_EVENT = 'candice:speech-drain';

/** Hard ceiling on phoneme timings per utterance (mirrors the native cap). */
const MAX_TIMINGS = 4096;

/** One phoneme span, exactly the WS-19 `PhonemeTiming` contract shape. */
export interface SpeechTiming {
  phoneme: string;
  startSec: number;
  endSec: number;
}

export interface SpeechStartPayload {
  schemaVersion: string;
  utteranceId: string;
  timings: SpeechTiming[];
}

export interface SpeechMarkerPayload {
  schemaVersion: string;
  utteranceId: string;
}

export interface SpeechTimingChannel {
  /** True while a native event listener is attached. */
  readonly listening: boolean;
  /** The utterance id the scheduler is currently animating, if any. */
  readonly activeUtteranceId: string | null;
  /** Detach all listeners and stop the scheduler. Idempotent. */
  dispose(): void;
}

/** Minimal event-listen contract; the Tauri event API satisfies it. */
export interface SpeechTimingListenApi {
  listen(event: string, handler: (event: { payload: unknown }) => void): Promise<() => void>;
}

export interface SpeechTimingChannelOptions {
  /** Injected for tests; defaults to the Tauri event API. */
  listenApi?: SpeechTimingListenApi;
}

function validUtteranceId(value: unknown): value is string {
  return typeof value === 'string'
    && value.length > 0
    && value.length <= 128
    && /^[!-~]+$/.test(value);
}

function validTiming(value: unknown): value is SpeechTiming {
  if (!value || typeof value !== 'object') return false;
  const timing = value as Record<string, unknown>;
  return typeof timing.phoneme === 'string'
    && timing.phoneme.length > 0
    && timing.phoneme.length <= 16
    && /^[\x21-\x7e]+$/.test(timing.phoneme)
    && typeof timing.startSec === 'number'
    && Number.isFinite(timing.startSec)
    && timing.startSec >= 0
    && typeof timing.endSec === 'number'
    && Number.isFinite(timing.endSec)
    && timing.endSec > timing.startSec;
}

/**
 * Parse a speech-start payload. Returns null for anything malformed —
 * the channel must never throw on real-world event traffic, and a
 * malformed payload must never reach the scheduler.
 */
export function parseSpeechStart(payload: unknown): SpeechStartPayload | null {
  if (!payload || typeof payload !== 'object') return null;
  const value = payload as Record<string, unknown>;
  if (value.schemaVersion !== SPEECH_TIMING_SCHEMA_VERSION
    || !validUtteranceId(value.utteranceId)
    || !Array.isArray(value.timings)
    || value.timings.length > MAX_TIMINGS) return null;
  const timings: SpeechTiming[] = [];
  for (const item of value.timings) {
    if (!validTiming(item)) return null;
    timings.push(item as SpeechTiming);
  }
  return { schemaVersion: value.schemaVersion, utteranceId: value.utteranceId, timings };
}

/** Parse a boundary/drain marker payload. Null for anything malformed. */
export function parseSpeechMarker(payload: unknown): SpeechMarkerPayload | null {
  if (!payload || typeof payload !== 'object') return null;
  const value = payload as Record<string, unknown>;
  if (value.schemaVersion !== SPEECH_TIMING_SCHEMA_VERSION
    || !validUtteranceId(value.utteranceId)) return null;
  return { schemaVersion: value.schemaVersion, utteranceId: value.utteranceId };
}

/**
 * Attach the native speech-timing events to a viseme scheduler.
 *
 * - `speech-start` loads the utterance's phoneme timings into the
 *   scheduler (the scheduler itself skips spans that do not map to a
 *   valid viseme event — garbage in, gap out, never a throw).
 * - `speech-boundary` stops the current utterance; a new start event
 *   follows if speech continues.
 * - `speech-drain` stops the current utterance; the mouth returns to
 *   closed. A drain for a stale utterance id is ignored.
 *
 * Returns a channel handle; `dispose()` detaches the listeners and
 * stops the scheduler. Safe as a no-op when the Tauri event API is
 * absent (plain-web boot path).
 */
export async function attachSpeechTimingChannel(
  scheduler: VisemeScheduler,
  options: SpeechTimingChannelOptions = {},
): Promise<SpeechTimingChannel> {
  let activeUtteranceId: string | null = null;
  let disposed = false;
  const unlisteners: Array<() => void> = [];

  const stopActive = (): void => {
    if (activeUtteranceId !== null) {
      scheduler.stop();
      activeUtteranceId = null;
    }
  };

  try {
    const listenApi: SpeechTimingListenApi = options.listenApi
      ?? (await import('@tauri-apps/api/event'));
    const unlistenStart = await listenApi.listen(SPEECH_START_EVENT, (event) => {
      if (disposed) return;
      const payload = parseSpeechStart(event.payload);
      if (!payload) return;
      // A new utterance replaces the current one; the scheduler's start
      // call resets its own state, so no explicit stop is needed here.
      scheduler.start(payload.timings);
      activeUtteranceId = payload.utteranceId;
    });
    const unlistenBoundary = await listenApi.listen(SPEECH_BOUNDARY_EVENT, (event) => {
      if (disposed) return;
      const payload = parseSpeechMarker(event.payload);
      if (!payload) return;
      if (activeUtteranceId !== null && payload.utteranceId !== activeUtteranceId) return;
      stopActive();
    });
    const unlistenDrain = await listenApi.listen(SPEECH_DRAIN_EVENT, (event) => {
      if (disposed) return;
      const payload = parseSpeechMarker(event.payload);
      if (!payload) return;
      if (activeUtteranceId !== null && payload.utteranceId !== activeUtteranceId) return;
      stopActive();
    });
    unlisteners.push(unlistenStart, unlistenBoundary, unlistenDrain);
  } catch {
    // Tauri event API absent (plain-web boot): the channel stays inert.
    // The scheduler remains idle-closed; the shell degrades to text.
  }

  return {
    get listening(): boolean {
      return !disposed && unlisteners.length > 0;
    },
    get activeUtteranceId(): string | null {
      return activeUtteranceId;
    },
    dispose(): void {
      if (disposed) return;
      disposed = true;
      for (const unlisten of unlisteners) {
        try { unlisten(); } catch { /* best-effort teardown */ }
      }
      unlisteners.length = 0;
      stopActive();
    },
  };
}

File: apps/candice-companion/src-tauri/src/speech_timing.rs
//! TTS speech-timing event surface (FIX-016).
//!
//! Minimal native emit in the audio duplex path: the TTS/duplex lanes
//! report the three timing facts this channel carries — speech start
//! (with phoneme timings), utterance boundary, and playback drain — and
//! the shell re-emits them as validated events on the existing bridge
//! event surface. The webview-side viseme scheduler (WS-12) consumes
//! those events; this module only validates payload shape before emit.
//!
//! No new auth surface: the events travel over the same `app.emit`
//! surface as `candice:bridge-question` / `candice:bridge-cancel`, the
//! commands live in the same `invoke_handler` list as the existing shell
//! commands, and no capability permission is added (`core:default`
//! already covers event emit/listen). Validation here is format
//! sanitation (finite spans, bounded count, opaque id), not
//! authorization — a webview caller is already inside the shell.

use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Emitter, Runtime};

pub const SPEECH_TIMING_SCHEMA_VERSION: &str = "1.0";
pub const SPEECH_START_EVENT: &str = "candice:speech-start";
pub const SPEECH_BOUNDARY_EVENT: &str = "candice:speech-boundary";
pub const SPEECH_DRAIN_EVENT: &str = "candice:speech-drain";

/// Hard ceiling on phoneme timings per utterance. Real Kokoro utterances
/// carry hundreds of spans; anything beyond this is rejected as malformed.
const MAX_TIMINGS: usize = 4096;

/// One phoneme span inside synthesized audio (mirror of the WS-19
/// `PhonemeTiming` contract in `src-tauri/tts/types.ts`).
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SpeechTiming {
    pub phoneme: String,
    pub start_sec: f64,
    pub end_sec: f64,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SpeechStartPayload {
    pub schema_version: String,
    pub utterance_id: String,
    pub timings: Vec<SpeechTiming>,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SpeechMarkerPayload {
    pub schema_version: String,
    pub utterance_id: String,
}

fn valid_utterance_id(value: &str) -> bool {
    !value.is_empty()
        && value.len() <= 128
        && value.bytes().all(|byte| matches!(byte, b'!'..=b'~'))
}

/// Is this string a plausible phoneme label?
///
/// THE authority for that question -- `engines.rs` calls this rather than
/// keeping its own copy. It used to keep one, and both copies independently
/// required `is_ascii_graphic()`. Kokoro drives espeak-ng, which emits IPA:
/// on a measured 48-span utterance that rule silently discarded 28 spans in
/// `filter_map`, and the 20 survivors were almost all consonants, so the
/// mouth stayed shut for 85% of the audio. Two agreeing copies of a wrong
/// rule look like corroboration; that is why there is now only one.
///
/// The rule is a DENY-list, not an allow-list, because the phoneme alphabet
/// belongs to the voice engine and will change again: reject control and
/// format characters (terminal escapes, bidi overrides, zero-width joiners)
/// and admit everything else. These strings are only ever used as viseme
/// lookup keys -- never rendered, never executed.
pub(crate) fn valid_phoneme(phoneme: &str) -> bool {
    !phoneme.is_empty()
        && phoneme.chars().count() <= 16
        && !phoneme.chars().any(|c| {
            c.is_control()
                || matches!(c,
                    '\u{200b}'..='\u{200f}'
                    | '\u{2028}'..='\u{202e}'
                    | '\u{2060}'..='\u{2064}'
                    | '\u{feff}')
        })
}

/// Validate one phoneme timing span. Mirrors the WS-12 ingestion rule:
/// non-finite or non-positive spans are garbage and must never reach the
/// scheduler.
fn valid_timing(timing: &SpeechTiming) -> bool {
    valid_phoneme(&timing.phoneme)
        && timing.start_sec.is_finite()
        && timing.end_sec.is_finite()
        && timing.end_sec > timing.start_sec
        && timing.start_sec >= 0.0
}

fn validate_utterance_id(value: &str) -> Result<(), String> {
    if valid_utterance_id(value) {
        Ok(())
    } else {
        Err("invalid utterance id".into())
    }
}

/// An utterance with zero timings is legal (the scheduler maps it to
/// idle); a burst beyond the cap is malformed input.
fn validate_timings(timings: &[SpeechTiming]) -> Result<(), String> {
    if timings.len() > MAX_TIMINGS {
        return Err("invalid timing count".into());
    }
    for timing in timings {
        if !valid_timing(timing) {
            return Err("invalid phoneme timing span".into());
        }
    }
    Ok(())
}

/// Emit speech start with phoneme timings. Also the seam a future native
/// duplex path calls directly (no webview round-trip).
pub fn emit_speech_start<R: Runtime>(
    app: &AppHandle<R>,
    utterance_id: &str,
    timings: &[SpeechTiming],
) -> tauri::Result<()> {
    app.emit(
        SPEECH_START_EVENT,
        SpeechStartPayload {
            schema_version: SPEECH_TIMING_SCHEMA_VERSION.into(),
            utterance_id: utterance_id.into(),
            timings: timings.to_vec(),
        },
    )
}

/// Emit an utterance boundary (the current utterance was replaced or
/// interrupted; a new start event follows if speech continues).
pub fn emit_speech_boundary<R: Runtime>(
    app: &AppHandle<R>,
    utterance_id: &str,
) -> tauri::Result<()> {
    app.emit(
        SPEECH_BOUNDARY_EVENT,
        SpeechMarkerPayload {
            schema_version: SPEECH_TIMING_SCHEMA_VERSION.into(),
            utterance_id: utterance_id.into(),
        },
    )
}

/// Emit playback drain (output provably silent; the mouth returns to
/// closed). The duplex controller's finish/tail path is the caller.
pub fn emit_speech_drain<R: Runtime>(
    app: &AppHandle<R>,
    utterance_id: &str,
) -> tauri::Result<()> {
    app.emit(
        SPEECH_DRAIN_EVENT,
        SpeechMarkerPayload {
            schema_version: SPEECH_TIMING_SCHEMA_VERSION.into(),
            utterance_id: utterance_id.into(),
        },
    )
}

#[tauri::command]
pub fn cmd_speech_timing_start<R: Runtime>(
    app: AppHandle<R>,
    utterance_id: String,
    timings: Vec<SpeechTiming>,
) -> Result<(), String> {
    validate_utterance_id(&utterance_id)?;
    validate_timings(&timings)?;
    emit_speech_start(&app, &utterance_id, &timings)
        .map_err(|error| format!("speech timing emit failed: {error}"))
}

#[tauri::command]
pub fn cmd_speech_timing_boundary<R: Runtime>(
    app: AppHandle<R>,
    utterance_id: String,
) -> Result<(), String> {
    validate_utterance_id(&utterance_id)?;
    emit_speech_boundary(&app, &utterance_id)
        .map_err(|error| format!("speech timing emit failed: {error}"))
}

#[tauri::command]
pub fn cmd_speech_timing_drain<R: Runtime>(
    app: AppHandle<R>,
    utterance_id: String,
) -> Result<(), String> {
    validate_utterance_id(&utterance_id)?;
    emit_speech_drain(&app, &utterance_id)
        .map_err(|error| format!("speech timing emit failed: {error}"))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn timing(phoneme: &str, start_sec: f64, end_sec: f64) -> SpeechTiming {
        SpeechTiming { phoneme: phoneme.into(), start_sec, end_sec }
    }

    #[test]
    fn accepts_valid_utterance_ids_and_rejects_control_or_oversized() {
        assert!(valid_utterance_id("engine-42"));
        assert!(valid_utterance_id("a"));
        assert!(!valid_utterance_id(""));
        assert!(!valid_utterance_id("bad\nid"));
        assert!(!valid_utterance_id(" "));
        assert!(!valid_utterance_id(&"x".repeat(129)));
    }

    #[test]
    fn timing_validation_mirrors_scheduler_ingestion_rules() {
        assert!(valid_timing(&timing("a", 0.1, 0.2)));
        assert!(valid_timing(&timing("aa", 0.0, 0.01)));
        assert!(!valid_timing(&timing("", 0.1, 0.2)));
        assert!(!valid_timing(&timing("a", 0.2, 0.1))); // non-positive span
        assert!(!valid_timing(&timing("a", f64::NAN, 0.1)));
        assert!(!valid_timing(&timing("a", 0.1, f64::INFINITY)));
        assert!(!valid_timing(&timing("a", -1.0, 0.1)));
        assert!(!valid_timing(&timing("a\tb", 0.1, 0.2)));
        // IPA is what the pinned voice actually emits. An ASCII-only rule
        // here silently thinned real utterances until the mouth stopped
        // moving; these are captured espeak-ng outputs, not invented.
        assert!(valid_timing(&timing("\u{259}", 0.1, 0.2)), "schwa");
        assert!(valid_timing(&timing("\u{e6}", 0.1, 0.2)), "ae");
        assert!(valid_timing(&timing("\u{14b}", 0.1, 0.2)), "eng");
        assert!(valid_timing(&timing("\u{2c8}", 0.1, 0.2)), "primary stress");
        assert!(valid_timing(&timing("\u{2d0}", 0.1, 0.2)), "length mark");
        assert!(valid_timing(&timing("\u{283}", 0.1, 0.2)), "esh");
        assert!(valid_timing(&timing(" ", 0.1, 0.2)), "word gap");
        // Control and format characters still do not get through.
        assert!(!valid_phoneme("a\u{1b}[31m"), "terminal escape");
        assert!(!valid_phoneme("a\u{202e}b"), "bidi override");
        assert!(!valid_phoneme("a\u{200b}b"), "zero width space");
        assert!(!valid_phoneme("a\u{feff}"), "byte order mark");
        assert!(!valid_phoneme(""), "empty");
        assert!(!valid_phoneme(&"\u{259}".repeat(17)), "over the 16-char cap");
        assert!(valid_phoneme(&"\u{259}".repeat(16)), "at the 16-char cap");
        assert!(!valid_timing(&timing("a", 0.1, 0.1)));
    }

    #[test]
    fn timing_lists_allow_empty_but_reject_oversized_bursts() {
        assert_eq!(validate_timings(&[]), Ok(()));
        assert!(validate_timings(&[timing("a", 0.1, 0.2)]).is_ok());
        let burst: Vec<SpeechTiming> = (0..=MAX_TIMINGS).map(|i| timing("a", i as f64, i as f64 + 0.5)).collect();
        assert!(validate_timings(&burst).is_err());
    }

    #[test]
    fn start_payload_serializes_camel_case_for_the_webview() {
        let payload = SpeechStartPayload {
            schema_version: SPEECH_TIMING_SCHEMA_VERSION.into(),
            utterance_id: "engine-42".into(),
            timings: vec![timing("a", 0.1, 0.2)],
        };
        let value = serde_json::to_value(&payload).expect("serializable");
        assert_eq!(value["schemaVersion"], "1.0");
        assert_eq!(value["utteranceId"], "engine-42");
        assert_eq!(value["timings"][0]["phoneme"], "a");
        assert_eq!(value["timings"][0]["startSec"], 0.1);
        assert!(value.get("schema_version").is_none());
    }

    #[test]
    fn marker_payload_serializes_camel_case() {
        let value = serde_json::to_value(SpeechMarkerPayload {
            schema_version: SPEECH_TIMING_SCHEMA_VERSION.into(),
            utterance_id: "engine-42".into(),
        }).expect("serializable");
        assert_eq!(value["schemaVersion"], "1.0");
        assert_eq!(value["utteranceId"], "engine-42");
    }
}
